### PR TITLE
Added dprint config and commit-on-master task

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
> 👉 _See [#65993 Using a formatter for formatting (instead of lint rules)](https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/65993#top) for context._

Corresponds to the changes described in https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/65993. This initial PR adds:

* `.github/workflows/format-and-commit.yml`: adds a workflow to run dprint and commit changes
  * Note that the trigger to run on `master` is commented out
* `.vscode/extensions.json`: recommends the dprint extension
* `.vscode/settings.template.json`: sets VS Code to format with the dprint extension by default
* `dprint.json`: configuration file for [dprint](https://dprint.dev), including `"excludes"` for (known or seemingly) auto-generated packages
* `README.md` docs around formatting
* `package.json`: a new dev dependency on `dprint`

~Because `format-and-commit.yml`'s commit trigger is commented out, merging this PR won't cause `dprint` to be applied to all files.~

Co-authored-by: @jakebailey

### CODE REVIEW

File: dprint.json
Comment (JoshuaKGoldberg) [reply dprint.json:16]: Ah good point. I'll put this in the discussion as a potential followup.

File: .github/workflows/format-and-commit.yml
Comment (JoshuaKGoldberg) [inline .github/workflows/format-and-commit.yml:1]: I'll send a corresponding PR to https://github.com/DefinitelyTyped/dt-mergebot with a discussion prompt on the strategy, then update this comment.

File: .github/workflows/format-and-commit.yml
Comment (jakebailey) [inline .github/workflows/format-and-commit.yml:18]: This is only going to work on branches of this PR (untrusted fork PRs do not get any perms that allow pushing), in which case I don't think there's a reason to have this workflow.

File: .github/workflows/format-check.yml
Comment (jakebailey) [inline .github/workflows/format-check.yml:7]: The status code is the thing to check here; I would just make this `dprint check` and that's it.

Comment (jakebailey) [inline .github/workflows/format-check.yml:7]: This needs an `npm install`, otherwise the version of dprint used will be unspecified which may cause problems if the plugins in the dprint config are too old to work with newer dprint versions.

File: .gitignore
Comment (jakebailey) [inline .gitignore:44]: I suspect we should write `!/.vscode/settings.json`, otherwise this actually allows people to commit settings at _any_ level.

File: .github/workflows/format-and-commit.yml
Comment (jakebailey) [inline .github/workflows/format-and-commit.yml:21]: It's `master` here still.

File: .github/workflows/format-and-commit.yml
Comment (jakebailey) [inline .github/workflows/format-and-commit.yml:16]: For this pipeline, we'll want to set up a concurrency rule which allows only the most recent run to succeed and cancels others in progress; that way multiple simultaneous commits won't both attempt to format and then fail.

That or we just schedule it.

File: .github/workflows/format-and-commit.yml
Comment (JoshuaKGoldberg) [reply .github/workflows/format-and-commit.yml:21]: ha, thanks..

**Comment by jakebailey** [comment]
Just as a meta-note about the workflows, I'd prefer if job configs were to come last; they get long and having them at the top will push info like the name/triggers/etc down to the bottom when it's usually important info. I don't think I've seen jobs stuck at the top before.

File: .github/workflows/format-check.yml
Comment (jakebailey) [inline .github/workflows/format-check.yml:7]: Is this working around the large text output?

At the end of it, we'd probably want to not hide all of the output text.

File: .github/workflows/format-check.yml
Comment (jakebailey) [reply .github/workflows/format-check.yml]: Note that the status thing was changed, but IIRC exit codes inside `$()` get swallowed?

File: .github/workflows/format-check.yml
Comment (JoshuaKGoldberg) [reply .github/workflows/format-check.yml:7]: Yes, right now it swallows all output and puts only a _"found ..."_ line in the step summary. What would we want the rest of the output for, though?

Amusingly, that line might be coincidentally logged from a file inside the `dprint check` output. But the step summary is only checked for existence right now in https://github.com/DefinitelyTyped/dt-mergebot/pull/475. So it wouldn't matter if it's the "right" message.

Comment (JoshuaKGoldberg) [reply .github/workflows/format-check.yml:7]: Removing this file per the _August 10th_ update in https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/65993.

File: .github/workflows/format-and-commit.yml
Comment (JoshuaKGoldberg) [reply .github/workflows/format-and-commit.yml]: Makes sense. I can switch this to the `push` trigger once it's ready to go.

File: .github/workflows/format-and-commit.yml
Comment (peterblazejewicz) [inline .github/workflows/format-and-commit.yml:24]:
```suggestion
        node-version: '16'
```
should this one be in sync with others? e.g.:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/.github/workflows/CI.yml#L15

Comment (peterblazejewicz) [inline .github/workflows/format-and-commit.yml:24]:
```suggestion
        node-version: '16'
```
should this one be in sync with others? e.g.:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/.github/workflows/CI.yml#L15

File: .github/workflows/format-and-commit.yml
Comment (peterblazejewicz) [inline .github/workflows/format-and-commit.yml:35]: > TODO: What to use for these?

we are setting up committer info that way in other place:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/.github/workflows/UpdateCodeowners.yml#L30-L31

File: .vscode/extensions.json
Comment (peterblazejewicz) [inline .vscode/extensions.json:1]: could be nice touch for community: add IDEAs family support:
https://plugins.jetbrains.com/plugin/18192-dprint
(not sure if it matches features head-to-head with VS though)

File: .github/workflows/format-and-commit.yml
Comment (JoshuaKGoldberg) [reply .github/workflows/format-and-commit.yml:34]: Perfect, thanks!

**Review by peterblazejewicz — APPROVED** [review]

File: .vscode/settings.json
Comment (jakebailey) [inline .vscode/settings.json:1]: Thinking about this, I'm 50/50 on whether or not this is a good idea for this repo; this will prevent people from modifying the file without it being dirty on disk, but I also think that it's not going to end up being used because it's a lot more ergonomic to open a specific types folder instead of the entire repo.

Comment (peterblazejewicz) [reply .vscode/settings.json:1]: I'm always doing `types/*`, never repo, that's a valid point

Comment (JoshuaKGoldberg) [reply .vscode/settings.json:1]: Interesting, I actually always do repo.

Comment (JoshuaKGoldberg) [reply .vscode/settings.json]: Changed to a `.vscode/settings.template.json`.

**Review by peterblazejewicz — APPROVED** [review]

File: .dprint.jsonc
Comment (jakebailey) [inline .dprint.jsonc:31]: This is duplicated.

File: .dprint.jsonc
Comment (jakebailey) [inline .dprint.jsonc:24]: This is dupliciated.

File: .dprint.jsonc
Comment (jakebailey) [inline .dprint.jsonc:19]: Can we be more specific for the exclusion on this one? IIRC it's just specific files; potentially they could have explicit dprint-ignore-files or something too.

Comment (JoshuaKGoldberg) [reply .dprint.jsonc:19]: Hmm, do you have a contact for someone who'd know for sure? I'm having a hard time determining which ones are auto-generated from just what's on disk or in PR history.

cc @peterblazejewicz based on recent PRs?

File: .vscode/extensions.json
Comment (JoshuaKGoldberg) [reply .vscode/extensions.json:1]: Hmm yeah I don't use any JetBrains/IDEAs projects myself so I'm not familiar. https://www.jetbrains.com/help/idea/managing-plugins.html#suggested-plugins seems like a nice idea. Maybe a good followup from someone who knows that ecosystem?

File: .dprint.jsonc
Comment (jakebailey) [reply .dprint.jsonc:19]: The file I'm thinking of is `inspector.d.ts` as it has an associated script in `scripts`.

Comment (JoshuaKGoldberg) [reply .dprint.jsonc]: Makes sense. I'll limit to just that.

**Review by peterblazejewicz — APPROVED** [review]

**Review by peterblazejewicz — APPROVED** [review]

**Comment by jakebailey** [comment]
Hush, bot

File: .github/workflows/format-and-commit.yml
Comment (jakebailey) [inline .github/workflows/format-and-commit.yml:14]: This will have to be `contents: write` eventually, right?

File: .github/workflows/format-and-commit.yml
Comment (jakebailey) [inline .github/workflows/format-and-commit.yml:34]:
```suggestion
        commit_author: "TypeScript Bot <typescriptbot@microsoft.com>"
```

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## tuya-panel-kit ([<kbd>unpkg</kbd>](https://unpkg.com/browse/tuya-panel-kit@latest/))
was missing the following properties:
1. Drawer
1. Wave
1. Diffusion
1. OfflineView
1. FullView
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 9e81ba2a2642451dbdebcd6868cd5be68b4bf664
</p>

File: .github/workflows/format-and-commit.yml
Comment (JoshuaKGoldberg) [reply .github/workflows/format-and-commit.yml:35]: If the bot is doing the merging it can always send a GitHub API request before doing the merge API request... yet again I am griping that GitHub doesn't expose a hook to do this upon clicking the green button

Comment (jakebailey) [reply .github/workflows/format-and-commit.yml]: The bot won't be doing the merging of these formatting PRs as they are pushed straight to `master`, so it's not gonna retitle these changes.

I'm not a fan of the `chore:` prefix or `npx` being mentioned; we don't use those styles for the current commits we make, and we probably won't be using `npx` forever either.

File: .github/workflows/format-and-commit.yml
Comment (jakebailey) [inline .github/workflows/format-and-commit.yml:20]: We may want to clone this with the bot PR; pushing with the regular GHA token will not allow GHA to run on the commit, which may mask errors caused by formatting modifying `ExpectType` and `// @ts-expect-error`. I'm not sure if we have that secret available here, though; I don't have permission to access the settings for this repo. @andrewbranch or @sandersn may have perms to check?

File: .github/workflows/format-and-commit.yml
Comment (JoshuaKGoldberg) [reply .github/workflows/format-and-commit.yml]: I really _really_ miss Codeflow's delineation between "resolved" and "closed"...

File: .github/workflows/format-and-commit.yml
Comment (andrewbranch) [reply .github/workflows/format-and-commit.yml:20]: There’s `secrets.GH_DT_MERGEBOT_TOKEN`

Comment (jakebailey) [reply .github/workflows/format-and-commit.yml:20]: Yeah, if that exists, we should revert `contents: write` and clone with that token instead. The only way that can go wrong is if we push unformatted code and cause a loop (which is why you can't do this normally).

File: .dprint.jsonc
Comment (jakebailey) [inline .dprint.jsonc:30]: Above here, also add `./types/google-publisher-tag`.

File: .github/workflows/format-and-commit.yml
Comment (jakebailey) [reply .github/workflows/format-and-commit.yml:20]:
```suggestion
    - uses: actions/checkout@v3
      with:
        token: ${{ secrets.GH_DT_MERGEBOT_TOKEN }}
```

File: README.md
Comment (jakebailey) [inline README.md:653]:
```suggestion
Any unformatted code will be automatically reformatted after being merged.
```

**Review by jakebailey — COMMENTED** [review]
I can't comment on the files, but we should also:

- Delete `.prettierrc.json`
- Replace the contents of `.prettierignore` with `*`, so that nobody accidentally runs it.

File: .github/workflows/format-and-commit.yml
Comment (jakebailey) [inline .github/workflows/format-and-commit.yml:14]:
```suggestion
  contents: read
```

File: .github/workflows/format-and-commit.yml
Comment (jakebailey) [inline .github/workflows/format-and-commit.yml:11]:
```suggestion
on:
  push:
    branches:
      - master
  workflow_dispatch: ~
```

File: .vscode/settings.template.json
Comment (jakebailey) [inline .vscode/settings.template.json:2]:
```suggestion
  "[typescript][typescriptreact][javascript][javascriptreact]": {
```

File: README.md
Comment (jakebailey) [inline README.md:372]: We should probably retain this example and mention `// dprint-ignore`.

File: README.md
Comment (jakebailey) [inline README.md:649]:
```suggestion
Yes, using [dprint](https://dprint.dev).
```

File: package.json
Comment (jakebailey) [inline package.json:43]:
```suggestion
        "dprint": "^0.41.0",
```

**Comment by jakebailey** [comment]
The only thing left is some sort of bot warning about not having formatted in a PR; will look into that tomorrow (probably just adding to the dangerfile).

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## lodash-es ([<kbd>unpkg</kbd>](https://unpkg.com/browse/lodash-es@latest/))
was missing the following properties:
1. commit
1. lodash
1. next
1. plant
1. thru

<details>
<summary>as well as these 13 other properties...</summary>
<p>toIterator, toJSON, value, valueOf, wrapperAt, wrapperChain, wrapperCommit, wrapperLodash, wrapperNext, wrapperPlant, wrapperReverse, wrapperToIterator, wrapperValue</p>
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against afb63cd6c6695f3ca6abed70c5dd981a2f60f53e
</p>

**Comment by jakebailey** [comment]
Updated to include a dangerbot block for unformatted added/modified files. It should show up in this PR eventually as this PR touches files that aren't formatted.

File: README.md
Comment (sandersn) [inline README.md:451]: this eventually needs more than a change of filenames.

Comment (jakebailey) [reply README.md:451]: Oops, I forgot to actually read the sentence; will fix.

Comment (jakebailey) [reply README.md:451]: Actually, is this not true? We're not allowing people to change formatting. Are you meaning that this should say how to opt-out for generated code?

Comment (sandersn) [reply README.md:451]: Yes, you can't change the settings. But it's more accurate to say that the whole repo is formatted with a tool, which is what the new paragraph says. Most of the rest either isn't relevant or isn't true anymore.

**Comment by jakebailey** [comment]
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/66235#issuecomment-1746123423 shows the new dangerbot output.

File: README.md
Comment (jakebailey) [reply README.md:451]: Yeah, I just didn't read this properly. The ending is the bad part. We have a completely different entry for formatting that says "we use dprint", so I'm just going to replace it with the new one so it gets more reading.

**Comment by jakebailey** [comment]
Okay, this PR should be ready to go, whenever we're ready to flip the switch! It'll immediately format about 141 files.

**Review by andrewbranch — APPROVED** [review]
👏

**Comment by jakebailey** [comment]
Ready to merge

**Comment by jakebailey** [comment]
image

**Pull request merged**
